File: scripts/bindings/ggplot_positional_label_functions.py
"""FIX-GG04 vector label functions through actual Python authoring and publication."""
import json,sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
registry=c.ExtensionRegistry.example();records=[]
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
cases=json.loads((ROOT/'fixtures/parity/ggplot2/positional-label-functions.json').read_text())['cases']
def data_for(t):return c.Data.columns({'x':c.column([v or 0. for v in t['inputs']],kind='float64').validity([v is not None for v in t['inputs']])},keys=list(range(100,100+len(t['inputs']))))
def layer():return c.points().name('marks')
def build(t,d):
 scale={'identity':c.scale_linear,'sqrt':c.scale_sqrt,'reverse':c.scale_reverse,'log10':lambda:c.scale_log(10.)}[t['transform']]()
 if t['limits']=='full':scale=scale.domain(1.,10.)
 axis=c.x_axis().scale(scale).range(100.,540.).guide_geometry({'labels':'Preserve'}).tick_format({'Registered':{'operation':{'id':'example.scale_labels','version':'1'},'parameters':t['label_mode']}})
 if t['break_mode']=='explicit':axis=axis.tick_values([-1,0,1,1,3,20,{'Number':{'number':'Infinity'}},{'Number':{'number':'NaN'}}])
 elif t['break_mode']=='empty':axis=axis.tick_values([])
 return c.plot(d).with_registry(registry).profile('Ggplot2_4_0_3').aes(c.aes().x('x').y(1.)).layer(layer()).x_axis(axis).y_axis(c.y_axis().visible(False)).build()
def check(t,frame):
 ticks=next(g for g in frame.guides()['guides'] if g['spec']['side']=='Bottom')['ticks']
 if all(isinstance(v,(float,int)) for v in t['result']['range']):
  expected=[label or '' for v,label in zip(t['result']['breaks'],t['result']['labels']) if isinstance(v,(float,int))]
  assert [tick['label'] for tick in ticks]==expected,(t,ticks,expected)
 return {'ticks':ticks}
for index,t in enumerate(cases):
 owned=[]
 try:
  d=data_for(t);owned.append(d);p=build(t,d);owned.append(p);wire=p.to_json()
  restored=c.Plot.from_json(wire,registry);owned.append(restored);assert restored.to_json()==wire
  for state in ('original','edited'):
   current=restored if state=='original' else restored.edit().layer('marks',layer()).build()
   if current is not restored:owned.append(current)
   request=output.request(current,options);owned.append(request);frame=request.prepare();owned.append(frame)
   assert 'error' not in t['result'],t
   records.append({'index':index,'state':state,**check(t,frame)})
   if state=='original' and index in (0,5,120,240,360):
    for fmt in ('svg','pdf','png'):(out/f'sample-{index}.{fmt}').write_bytes(frame.export(fmt))
 except c.ChartError as error:
  assert 'error' in t['result'],(t,str(error));assert error.code=='CHART_SCHEMA_CONFLICT',(t,error.code)
  records.append({'index':index,'error':error.code})
 finally:
  for obj in reversed(owned):obj.dispose()
assert len(records)==668,len(records)
for transform in ('identity','sqrt','log10','reverse'):
 for mode in ('indexed','missing'):
  selected={t['population']:t for t in cases if t['transform']==transform and t['limits']=='full' and t['break_mode']=='automatic' and t['label_mode']==mode};owned=[]
  try:
   original=data_for(selected['ordinary']);owned.append(original);p=build(selected['ordinary'],original);owned.append(p);chart=p.chart();owned.append(chart);wire=p.to_json();request=output.request(p,options);owned.append(request);held=request.prepare();owned.append(held);scene=held.scene()
   for population in ('constant','missing','all_missing','empty','ordinary'):
    t=selected[population];replacement=data_for(t);owned.append(replacement);tx=chart.transaction();owned.append(tx);builder=tx.replace(original,replacement);owned.append(builder);update=builder.build();owned.append(update);assert 'Applied' in chart.commit(update)
    fresh=build(t,replacement);owned.append(fresh);batch=fresh.chart();owned.append(batch)
    a=chart.semantics();b=batch.semantics();actual={'limits':a.get('positional_limits'),'domains':a['layers'][0]['domains']};assert actual=={'limits':b.get('positional_limits'),'domains':b['layers'][0]['domains']};assert held.scene()==scene and p.to_json()==wire
    records.append({'transform':transform,'mode':mode,'replacement':population,'semantics':actual})
  finally:
   for obj in reversed(owned):obj.dispose()
assert len(records)==708
(out/'records.json').write_text(json.dumps(records,indent=2));options.dispose();output.dispose();registry.dispose()
print('PASS Python: 708 positional label states and 15 publications.')

File: crates/chart-core/tests/ggplot_positional_label_functions.rs
//! FIX-GG04: registered label vectors against pinned source and guide-key results.
use chart_core::{
    ChartResult, GuideId, Rect, ResourceId, Revision,
    composition::ScaleValue,
    grammar::{CustomGuideFormatter, ExtensionDescriptor, ExtensionRegistry, GuideLabelsInput},
    layout::*,
    prelude::*,
    services::*,
};
use serde_json::{Value, json};
use std::sync::{Arc, Mutex};
struct Formatter(Arc<Mutex<Vec<Value>>>, bool);
impl CustomGuideFormatter for Formatter {
    fn descriptor(&self) -> ExtensionDescriptor {
        ExtensionDescriptor::batch("test.scale_labels", Revision::new(1), self.1)
    }
    fn validate(&self, _: &Value) -> ChartResult<()> {
        Ok(())
    }
    fn format_labels(&self, input: GuideLabelsInput<'_>) -> ChartResult<Vec<Option<String>>> {
        let values = input
            .values
            .iter()
            .map(|v| match v {
                ScaleValue::Number(v) if v.is_nan() => Value::Null,
                ScaleValue::Number(v) if v.is_infinite() => {
                    json!(if *v > 0. { "Infinity" } else { "-Infinity" })
                }
                ScaleValue::Number(v) => json!(v),
                ScaleValue::Category(v) => json!(v),
                ScaleValue::MissingCategory => Value::Null,
                _ => panic!("unexpected timestamp"),
            })
            .collect::<Vec<_>>();
        self.0
            .lock()
            .unwrap()
            .push(json!({"values":values,"names":input.names.unwrap_or(&[])}));
        let mut labels = if input.values.is_empty() {
            vec![Some("/0".into())]
        } else {
            (1..=input.values.len())
                .map(|i| Some(format!("{i}/{}", input.values.len())))
                .collect::<Vec<_>>()
        };
        match input.parameters.as_str().unwrap() {
            "missing" => {
                for (i, label) in labels.iter_mut().enumerate() {
                    if i % 2 == 1 {
                        *label = None;
                    }
                }
            }
            "short" => labels.truncate(1),
            "empty" => labels.clear(),
            "indexed" | "named" => {}
            _ => panic!("unexpected mode"),
        }
        Ok(labels)
    }
}

struct Metrics;
impl TextMeasurer for Metrics {
    fn measure(&self, r: TextRequest<'_>) -> ChartResult<TextMetrics> {
        TextMetrics::new(r.text.len() as f64 * 5., 8., 2.)
    }
}
#[test]
fn positional_numeric_labels_match_reference_candidates_and_visible_ticks() {
    let fixture: Value = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/positional-label-functions.json"
    ))
    .unwrap();
    let mut passed = 0;
    for (index, c) in fixture["cases"].as_array().unwrap().iter().enumerate() {
        let calls = Arc::new(Mutex::new(vec![]));
        let mut registry = ExtensionRegistry::default();
        registry
            .register_guide_formatter(Arc::new(Formatter(calls.clone(), true)))
            .unwrap();
        let inputs = c["inputs"].as_array().unwrap();
        let data = Data::columns()
            .column(
                "x",
                chart_core::plot::column(
                    inputs
                        .iter()
                        .map(|v| v.as_f64().unwrap_or(0.))
                        .collect::<Vec<_>>(),
                )
                .validity(inputs.iter().map(|v| !v.is_null()).collect()),
            )
            .column("y", vec![1.; inputs.len()])
            .build()
            .unwrap();
        let mut scale = match c["transform"].as_str().unwrap() {
            "identity" => scale_linear(),
            "sqrt" => scale_sqrt(),
            "log10" => scale_log(10.),
            "reverse" => scale_reverse(),
            _ => unreachable!(),
        };
        if c["limits"] == "full" {
            scale = scale.domain(1., 10.);
        }
        let mut axis = x_axis()
            .scale(scale)
            .tick_format(Some(GuideFormatter::Registered {
                operation: chart_core::grammar::OperationRef::new(
                    "test.scale_labels",
                    Revision::new(1),
                ),
                parameters: c["label_mode"].clone(),
            }))
            .guide_geometry(Some(GuideGeometry {
                labels: Some(GuideLabelPolicy::Preserve),
                ..Default::default()
            }));
        if c["break_mode"] == "explicit" {
            axis = axis.tick_values(Some(
                vec![-1., 0., 1., 1., 3., 20., f64::INFINITY, f64::NAN]
                    .into_iter()
                    .map(ScaleValue::Number)
                    .collect(),
            ));
        } else if c["break_mode"] == "empty" {
            axis = axis.tick_values(Some(vec![]));
        }
        let result = (|| -> ChartResult<_> {
            let plot = plot(data)
                .profile(Profile::Ggplot2_4_0_3)
                .aes(aes().x("x").y("y"))
                .layer(points())
                .x_axis(axis)
                .extensions(Arc::new(registry))
                .build()?;
            let prepared = plot.chart()?.prepare()?;
            layout(
                prepared,
                &LayoutRequest::new(
                    Rect::new(0., 0., 640., 360.)?,
                    Units::LogicalPixels,
                    ResourceDescriptor {
                        id: ResourceId::new(1),
                        revision: Revision::INITIAL,
                        kind: ResourceKind::Font,
                        byte_len: 1,
                    },
                ),
                &Metrics,
            )
        })();
        assert_eq!(
            result.is_ok(),
            c["result"].get("error").is_none(),
            "case {index}: {c} got {:?}",
            result.as_ref().err()
        );
        let actual = calls.lock().unwrap();
        assert!(!actual.is_empty(), "case {index}: callback skipped, {c}");
        for call in actual.iter() {
            let expected = &c["calls"][0];
            assert_eq!(call["names"], expected["names"], "case {index}");
            let a = call["values"].as_array().unwrap();
            let b = expected["values"].as_array().unwrap();
            assert_eq!(a.len(), b.len(), "case {index}: {a:?} != {b:?}");
            for (a, b) in a.iter().zip(b) {
                if let (Some(a), Some(b)) = (a.as_f64(), b.as_f64()) {
                    assert!(
                        (a - b).abs() <= 3e-12 * b.abs().max(1.),
                        "case {index}: {a} != {b}"
                    );
                } else {
                    assert_eq!(a, b, "case {index}");
                }
            }
        }
        if let Ok(frame) = result {
            passed += 1;
            let finite_range = c["result"]["range"]
                .as_array()
                .unwrap()
                .iter()
                .all(Value::is_number);
            if finite_range {
                let labels = c["result"]["breaks"]
                    .as_array()
                    .unwrap()
                    .iter()
                    .zip(c["result"]["labels"].as_array().unwrap())
                    .filter(|(v, _)| v.is_number())
                    .map(|(_, v)| v.as_str().unwrap_or(""))
                    .collect::<Vec<_>>();
                let got = frame.guides()[&GuideId::new(0)]
                    .ticks
                    .iter()
                    .map(|t| t.label.as_str())
                    .collect::<Vec<_>>();
                assert_eq!(got, labels, "case {index}");
            }
        }
    }
    assert_eq!(passed, 188);
}
